### PR TITLE
chore: split export dependencies into respective framework groups

### PR DESCRIPTION
**Pull request by Gornoka**
### TLDR: No longer install 2 Gigs of TF deps when all you want is exporting to onnx.

This PR splits the dependencies for the different ML export Frameworks into different extra groups and adds a meta group for backwards compatibility, This simplifies downstream handling of the library, when only requiring specific exports, especially the TF and TF-lite related dependencies have made it annoying to install Those dependencies.

When using the previous dependencies with tools like UV installation of the export extra was not straightforward on Windows and maybe other systems as well, as `tensorflow-decision-forests` is not supported on windows. (There are Workarounds with OS specific version pinning, but those are cumbersome and fragile).

As The dependency trees of the different frameworks are mostly disjunct, besides some common reliance on ONNX This seems like a nice solution to me, where you get very decent behavior and dependency trees for every export use-case (or the old Linux ~only "export all at once" behavior if you choose to use the old extra).

This also adds the magic 'check_requirements' dependencies from the TFexporter directly to the pyproject.toml, which gives end-users more transparency about the things they are installing.

I did not add the dependencies for the completely hidden exporters (those which were not part of the old export requirements at all), if you want me to add them here as well I can do that.

    I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Refactors and organizes export-related dependencies into separate groups for ONNX, OpenVINO, TensorFlow, and CoreML, making export options more modular and manageable. 🛠️✨

### 📊 Key Changes
- Splits export dependencies into distinct groups: `export-onnx`, `export-openvino`, `export-tensorflow`, and `export-coreml`.
- Updates the main `export` group to include all these new sub-groups.
- Adds and clarifies several TensorFlow-related dependencies for improved compatibility and functionality.
- Improves platform and Python version checks for certain dependencies.

### 🎯 Purpose & Impact
- **Easier Installation:** Users can now install only the export dependencies they need, reducing unnecessary packages and potential conflicts.
- **Better Compatibility:** Fine-tuned dependency management ensures smoother exports across different platforms and Python versions.
- **Improved Maintainability:** Clearer structure makes it easier for developers to update or troubleshoot export dependencies in the future.

This update streamlines the export process, making it more flexible and reliable for all users! 🚀

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Gornoka, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline the review process—a member of the Ultralytics engineering team will assist you further soon.

Please review the following checklist to ensure your PR is ready for integration:

- ✅ **Define a Purpose**: Clearly explain the motivation and scope of your changes in the PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear and follow the project's conventions.
- ✅ **Synchronize with Source**: Make sure your PR is up-to-date with the `main` branch of `ultralytics/ultralytics`. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Confirm all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. Please address any failures if they occur.
- ✅ **Update Documentation**: Update the [documentation](https://docs.ultralytics.com/) for any new or modified features, especially regarding the new export dependency groups.
- ✅ **Add Tests**: If applicable, include or update tests that cover your changes. Verify that all tests run successfully.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message (which you have already done—thank you!).
- ✅ **Minimize Changes**: Limit your changes to the **minimum** required for this update. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more details, please consult our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If you have any questions, feel free to comment here. Thank you for helping improve Ultralytics! 🛠️✨

**Comment by Laughing-q** [comment]
@Gornoka The idea is nice! I think the simplest way there is to directly remove all the tensorflow deps in the `export` section here, since we are handling all the deps here: https://github.com/ultralytics/ultralytics/blob/d564179793b6e6579a08351e2a7d948fe19ad8ca/ultralytics/engine/exporter.py#L964-L977, this way users only install some basic packages like onnx by `pip install ultralytics[export]`, and not bothered by tensorflow installation overhead. The tensorflow installation would occur only when they are exporting relevant formats.

**Comment by Gornoka** [comment]
> @Gornoka The idea is nice! I think the simplest way there is to directly remove all the tensorflow deps in the `export` section here, since we are handling all the deps here:
> 
> https://github.com/ultralytics/ultralytics/blob/d564179793b6e6579a08351e2a7d948fe19ad8ca/ultralytics/engine/exporter.py#L964-L977
> 
> , this way users only install some basic packages like onnx by `pip install ultralytics[export]`, and not bothered by tensorflow installation overhead. The tensorflow installation would occur only when they are exporting relevant formats.

I strongly dislike the idea of more reliance on runtime dependency installation, as it is a nightmare to audit these dependencies, almost impossible to trace their state and it significantly complicates usage in container environments.

Regardless I don't think this pr should touch this behaviour, as it would be a breaking change.

**Comment by Gornoka** [comment]
Are there any more steps required to proceed with this ?

**Comment by Laughing-q** [comment]
@Gornoka yeah this is a good idea. Sometimes `check_requirements` does not work very well when there's multiple different exporting tasks in a sequence running(assuming users don't install all the dependencies beforehand), for example our ci tests is install most of the needed dependencies by `check_requirements` but sometimes there's conflicts between package installation and cause tests fail.

By this PR we could install all the dependencies by `pip install -e ".[export]"`, which is great. Meanwhile we also want to keep `check_requirements` in exporter as this automate installation is also very convenient for users to do a quick export for a specific format without concerning the dependency. 
Hence this way there would be two places we need to maintain all the dependency versions, `pyproject.toml` and all the `check_requirements`, which we don't really want. 
I'm planning to revisit this later this week trying to figure out a solution cover both sides but no extra maintenance cost, or do you have some insights on this?

**Comment by Gornoka** [comment]
### Extra split related things
This PR mostly splits the old `export` group into multiple groups, This does not create any additional maintenance overhead, as the dependencies were already in 2 locations, it just allows users to install less dependencies if they explicitly want to.

There is one other thing I did which technically may increase the overhead, I added the missing runtime installed dependencies from TF directly to the TF group of the extras. 
From my point of view it was a bug that they were missing in the `export` group before, as most of them were already in that group.

old TF export related packes in `export`
````
    "tensorflow>=2.0.0", # TF bug https://github.com/ultralytics/ultralytics/issues/5161
    "tensorflowjs>=2.0.0", # TF.js export, automatically installs tensorflow
    "tensorstore>=0.1.63; platform_machine == 'aarch64' and python_version >= '3.9'", # for TF Raspberry Pi exports
    "h5py!=3.11.0; platform_machine == 'aarch64'", # fix h5py build issues due to missing aarch64 wheels in 3.11 release
````

old tf related packages that were missing in `export` ( but found in runtime install )
````
    "tf_keras",  # required by 'onnx2tf' package
    "sng4onnx>=1.0.1",  # required by 'onnx2tf' package
    "onnx_graphsurgeon>=0.3.26",  # required by 'onnx2tf' package
    "ai-edge-litert>=1.2.0; python_version >='3.9' and platform_system != 'Windows'",  # required by 'onnx2tf' package, only available for python >= 3.9
    "onnx2tf>=1.26.3; python_version >='3.10'", # below 3.10 onnx2tf may not be resolvable on windows,
    # it will try to resolve this on runtime
    "onnxslim>=0.1.53",
    "onnxruntime", # in case a gpu is available the gpu version will be installed at runtime
    "protobuf>=5",
````

If we were to remove the second group from the new `export-tensorflow` extra the maintainability would not change at all in this PR, and the old ( imho broken) behaviour would stay exactly the same. 
The old behaviour also means that when you install the extra for tf install you never have all dependencies required for that export and will always rely on partial runtime installation.


### Runtime install related things
I understand that you prefer the automagic feeling and want to install it directly for your users.

What you could do is to refer your runtime install to the extra dependencies in your pyproject.toml, so that both installation methods use the same dependencies.

You could do so by parsing the pyproject.toml file of your installation at runtime, and instead of providing a list of additional dependencies at 10 locations in your code, you could check which extra is required for this operation (this will require including your pyproject.toml in the installation). That way you would eliminate the duplication in your code and enable automatic dependency updates for your extras with tools like dependabot.

I don't think it makes sense to add a new runtime installation mechanic into this PR as they are independent features.

File: pyproject.toml [hunk 1]
```diff
@@ -91,15 +91,36 @@ dev = [
...
+    "onnxslim>=0.1.53",
+    "onnxruntime", # in case a gpu is available the gpu version will be installed at runtime
+    "protobuf>=5",
+]
+
+export-coreml = [
+    "coremltools>=8.0; platform_system != 'Windows' and python_version <= '3.13'", # CoreML supported on macOS and Linux
+    "scikit-learn>=1.3.2; platform_system != 'Windows' and python_version <= '3.13'", # CoreML k-means quantization
+]
+export = [
```
Comment (glenn-jocher) [inline pyproject.toml:121]: I don't understand. Won't this use the pypi version of `ultralytics` instead of the local version? This won't work for development naturally.

**Comment by glenn-jocher** [comment]
@Gornoka also you don't need to preinstall onnx, it will install automatically if your environment doesn't have it upon export, i.e. you can just run this and everything will work:

```
pip install ultralytics

yolo export format=onnx
```

File: pyproject.toml (hunk 1 above)
Comment (Gornoka) [reply pyproject.toml:121]: my understanding of python packaging extras and my testing does not seem to implicate that. As it installs `tf-keras` and `sng4onnx` (which is not part of that extra on main) when installing the `export` extra on my branch.


````
uv pip install  .[export]     
Using Python 3.12.4 environment at: .venv_3.12
Resolved 79 packages in 105ms
Uninstalled 2 packages in 3ms
Installed 42 packages in 2.51s
 + absl-py==2.3.0
 + astunparse==1.6.3
 + coloredlogs==15.0.1
 + flatbuffers==25.2.10
 + gast==0.6.0
 + google-pasta==0.2.0
 + grpcio==1.72.1
 + h5py==3.13.0
 + humanfriendly==10.0
 + keras==3.10.0
 + libclang==18.1.1
 + markdown==3.8
 + markdown-it-py==3.0.0
 + mdurl==0.1.2
 + ml-dtypes==0.5.1
 + namex==0.1.0
 + onnx==1.17.0
 + onnx-graphsurgeon==0.5.8
 + onnx2tf==1.27.10
 + onnxruntime==1.22.0
 + onnxslim==0.1.56
 + openvino==2025.1.0
 + openvino-telemetry==2025.1.0
 + opt-einsum==3.4.0
 + optree==0.16.0
 - packaging==23.2
 + packaging==20.9
 + protobuf==5.29.5
 + pygments==2.19.1
 + pyreadline3==3.5.4
 + rich==14.0.0
 + sng4onnx==1.0.4
 + tensorboard==2.19.0
 + tensorboard-data-server==0.7.2
 + tensorflow==2.19.0
 + tensorflow-hub==0.12.0
 + tensorflowjs==3.18.0
 + termcolor==3.1.0
 + tf-keras==2.19.0
 ~ ultralytics==8.3.150 (from file:///C:/git/ultralytics)
 + werkzeug==3.1.3
 + wheel==0.45.1
 + wrapt==1.17.2

````

pip install . and pip install -e . appeared to also process this correctly.

**Comment by Gornoka** [comment]
> @Gornoka also you don't need to preinstall onnx, it will install automatically if your environment doesn't have it upon export, i.e. you can just run this and everything will work:
> 
> ```
> pip install ultralytics
> 
> yolo export format=onnx
> ```

if possible , I want to avoid automatic installation at runtime using the ("YOLO_AUTOINSTALL") variable. But that is not really an option if some features only work with autoinstall enabled.

**Comment by glenn-jocher** [comment]
@Gornoka thanks for the thoughtful PR and detailed explanations! Your approach of splitting export dependencies into modular groups makes a lot of sense, especially for containerized environments and dependency auditing.

The idea of parsing pyproject.toml at runtime to eliminate duplication between extras and `check_requirements` is clever - that would solve the maintenance overhead concern while keeping both installation methods available. We could explore that as a follow-up enhancement.

For now, this PR provides immediate value by giving users granular control over export dependencies without breaking existing workflows. Let's move forward with this approach.

**Comment by Gornoka** [comment]
If anything besides the rebase need to be done let me know, but I will not be able to do it before next tuesday.

**Comment by glenn-jocher** [comment]
@Gornoka looks good! The rebase should be sufficient. Once that's done we'll review and merge. The modular approach you've implemented strikes the right balance between user flexibility and maintaining backward compatibility. Take your time with the rebase - we can wait until after Tuesday.

**Comment by Gornoka** [comment]
> @Gornoka looks good! The rebase should be sufficient. Once that's done we'll review and merge. The modular approach you've implemented strikes the right balance between user flexibility and maintaining backward compatibility. Take your time with the rebase - we can wait until after Tuesday.


rebase is done

**Comment by glenn-jocher** [comment]
@Gornoka perfect! Thanks for completing the rebase. We'll review the changes and get this merged. The modular dependency structure you've implemented will be a valuable improvement for users who need specific export formats without the overhead of unnecessary dependencies.

**Comment by Gornoka** [comment]
@glenn-jocher is there anything else from my side needed to advance this PR ?

**Comment by glenn-jocher** [comment]
@Gornoka no, everything looks good from your side! The PR is ready for our final review and merge. Thanks for your patience and the excellent work on modularizing the export dependencies - this will definitely benefit users who want to avoid the TensorFlow overhead when only needing ONNX exports.

**Comment by Gornoka** [comment]
hi @glenn-jocher sorry for the necromancy but it appears that this issue got buried a little, if you need anything from my side to advance the split please let me know.
If you should not want to merge a change like this that is obviously also fine, but please let me know if that is the case.

**Comment by glenn-jocher** [comment]
Thanks for the nudge—no further changes needed on your side; I’ve re-queued this for maintainer review and will clear the stale label, keeping your modular extras as proposed and `check_requirements` as a fallback—if minor tweaks (CI or docs) are needed, we’ll push them directly to your branch.

**Comment by Gornoka** [comment]
Sorry to bother you again with more necromancy @glenn-jocher .
On another note nice round you got there.

**Comment by glenn-jocher** [comment]
No bother at all—and thanks, that’s really down to the Ultralytics team and YOLO community; your PR is still in our review queue and there’s nothing else needed from your side right now.

**Comment by Gornoka** [comment]
@onuralpszr could you please elaborate why this was closed ?

**Comment by onuralpszr** [comment]
> @onuralpszr could you please elaborate why this was closed ?

Hello @Gornoka, sorry for late answer (I forgat sent message) 

It was a very old PR and I didn't see any activity so I closed based on that plus we had auto installer to handle those packages. We also kept this separation because some of the packages cause conflict in their packages so we didn't include or include when we need CI or if it is no conflict then we consider those cases.

**Comment by Gornoka** [comment]
> > @onuralpszr could you please elaborate why this was closed ?
> 
> Hello @Gornoka, sorry for late answer (I forgat sent message)
> 
> It was a very old PR and I didn't see any activity so I closed based on that plus we had auto installer to handle those packages. We also kept this separation because some of the packages cause conflict in their packages so we didn't include or include when we need CI or if it is no conflict then we consider those cases.

The inactivity was due to waiting for review, having hidden dependencies is still not great from a security and auditability perspective, exclusive dependencies could be specified for newer versions of uv, I doubt that pip will get similar features soon. 

I had opened a new pr for the rebased branch as i assumed it was just an automatic close ( #23929 ).  If that is not helpfull you can close that as well.
